### ISSUE TITLE
Add TQ as a TSV

### ISSUE THREAD

**Issue by birchamp** [unfoldingWord/tc-create-app#409]
Coming out of the spike #360

This will require different display columns than the TSV for TNs. Note that the new TSVs can have references with verse ranges.
The goal is to show the MD for the questions using the data in the TSV to index the questions according to the verse.  
 -  MD should be displayed the way TN displays but only display default column that are not empty.  
 -  The `<br>` to separate the question and answer should be separate lines with whitespace between them. 

It will necessary to work with @RobH123 on learning the new TSV format.

**Comment by birchamp** [comment]
@ancientTexts-net most fields should be hidden by default.

**Comment by mandolyte** [comment]
Working branches for new TSV format are always named "newFormat".

Here is the manifest entry for Titus:
```yaml

-
--
  | title: 'Titus'
  | versification: 'ufw'
  | identifier: 'tit'
  | sort: 56
  | path: './TIT_tq.tsv'
  | categories: [ 'bible-nt' ]
```

Link to Titus: https://qa.door43.org/unfoldingWord/en_tq/src/branch/newFormat/TIT_tq.tsv

**Comment by mandolyte** [comment]
Here is an example of the annotation:
```
What position and responsibility does an overseer have in God’s household?\n\n> He is like God’s household manager.
```

Here is what it should look like in preview mode:

What position and responsibility does an overseer have in God’s household?

> He is like God’s household manager.

**Comment by mandolyte** [comment]
@BincyJ @birchamp I have expanded on the requirements and have tagged others for confirmation for parts of it.

The TQ TSV format only uses three columns:
- Reference
- ID
- Annotation
(@RobH123 -- true?)


So in tc-create, only three columns should be available. Alternatively, if they are available, then it should be an error for them to have content. (@RobH123 -- true?)

The annotation columns is a specialized form of Markdown. The differences and conventions are:
- The question and answer are in the same Markdown.
- The question is followed by 5 characters `\n\n>`. The first two characters `\n` indicate a newline that ends the question. The second two characters `\n` indicates a blank line that should separate the question and answer. The fifth character indicates the beginning of a Markdown block quote.
- The above string must be pre-processed for preview:
    - The question and answer must be separated by a blank line
    - The answer must be treated as an HTML block quote (indented lines of text with a left vertical bar)


**Possible Issues**

1. The bi-directional translation between HTML and Markdown must be updated to change `\n` to/from `<br>`. (@ancientTexts-net -- FYI)
2. The blockquote visual aid (vertical bar) does not work in currently. It only indents. (@birchamp is this a problem?)
3. The answer must be written in Markdown mode since there is no way to indicate a block quote in the HTML preview mode. (@birchamp @BincyJ -- this sort of goes against user preference for editing in HTML mode)
4. Does the CV package handle tQ TSV? And if so, does it generate a notice if columns Tags, SupportReference, Quote, or Occurrence have content? (@RobH123 -- need your input)

**Comment by RobH123** [comment]
Ideally the `\n` should be converted to a newline in the new TSV reading function. See https://en.wikipedia.org/wiki/Tab-separated_values#Conventions_for_lossless_conversion_to_TSV

(Putting in <br> or whatever ideally is part of md -> HTML conversion.)

**Comment by RobH123** [comment]
I would suggest something in a log @mandolyte if there's something in an unexpected column and then ignore it. It shouldn't happen (with present data), however, we don't want it to fail if someone puts something in there. (Remember there's lots of ways for humans to edit / mess-up files.)

But in general, I think most of your assumptions above are correct.

**Comment by mandolyte** [comment]
@birchamp @RobH123 Quick question... currently the new TSV format for tQ files is not saved in master branch. They are in a branch named "newFormat". My question is whether I can assume that the new format branch will be merged into master and that is where tc-create should look for them.

**Comment by mandolyte** [comment]
@klappy Wondered if you have any advice on the following findings...

Whereas I have the 7 column TSV version for tQ working, I have not been successful in making the round trip from file -> to md in the datatable -> to html ->  (some edits) -> to md -> saved file.

I have found a number of places that all seem to impact the md from/to html (and perhaps back to DCS?). They are:
- datatable-translatable has input and output filters
- markdown-translatable has entire TS package `markdown-converter.ts` that has a slew of functions and filters (I even found one that ostensibly is doing the same thing I had down with filters to remove zero-width-spaces from beginning and end of the cell value)

My last iteration today found that what is saved in DCS is NOT what I see in the raw markdown. So evidently there is some other interference going on - which led me to the markdown converter package. I'm guessing that what I tried to do in the datatable filters is being altered.

Is there a good reason to make changes in both places? Should I stick with the markdown converter package?

My plan for tomorrow is to study the markdown converter package.
@birchamp @ancientTexts-net

**Comment by klappy** [comment]
Ideally we need to identify the proper separation of concerns. Where should the filter exist? Is it something that the UI cares about? Or is it something that the outer layers interacting with reading and saving the file should deal with? What tests can we create and at what point is it valuable for us to know that this is handled? When we write the tests that fail, then we can ensure it is handled properly.

**Comment by mandolyte** [comment]
:-) I'll take that as encouragement to study how that part of the code actually works. So far, I've just been guessing, testing, and failing.

For posterity, my notes will be [here](https://git.door43.org/cecil.new/journals/src/branch/master/DataTable-Markdown-translatable.md)

**Comment by mandolyte** [comment]
@birchamp So I had some thoughts about https://github.com/unfoldingWord/tc-create-app/issues/621 while trudging thru the code... Perhaps that would be a nice research project for the create 2.0 team. Some ideas:
1. Is only active, one instance at a time. Why? this gets around the problem noted before that you will run out of memory quickly if hundreds of instances are running.
2. The above implies that the user would have indicate intent to edit, say via the standard "pencil" icon or an on blur event.
3. Research how to replicate how section alignment can be done (or perhaps evaluate whether needed).

**Comment by mandolyte** [comment]
This issue can be unblocked now. The next few posts here will be to document what the new spec requires us to implement.

The new spec is [here](https://forum.door43.org/t/draft-parascriptural-tab-separated-value-format-specification-v2/502)

**Comment by mandolyte** [comment]
This post has the general requirements that govern all TSV formats
- [ ] The new TSV repos will include the OBS material. They will be treated like another bible book with the id "OBS".
- [ ] The filenames in the repos will be of the form BBB_xx.tsv, where 'BBB' is the book id (including OBS) and xx is the two letter abbreviation of the resource type.
- [ ] The resource abbreviations are: tn, tq, sn, sq, and twl.
- [ ] The repos will be named by language id and resource id. For example, en_tn, en_tq, en_sn, en_sq, and en_twl
- [ ] Note that the files will be listed in DCS in alphabetical order, since the numeric prefix for biblical order is no longer used.
- [ ] The TSV spec we will be implementing is described [here](https://en.wikipedia.org/wiki/Tab-separated_values#Conventions_for_lossless_conversion_to_TSV). And is defined [here](https://www.iana.org/assignments/media-types/text/tab-separated-values)
- [ ] The above spec requires that each line have the same number of tabs. *Should we consider re-laxing this requirement?*
- [ ] While not part of the IANA spec, there are several generally accepted escape sequences which we will support. These are: `\n, \t, \r, and \\`

**Comment by mandolyte** [comment]
For TQ and SQ:

These TSV files have FIVE columns: Reference, ID, Tags, Question, Response.

Since these are identical, one set of changes will work for both.

**Comment by mandolyte** [comment]
These changes cannot be deployed into production until:

1. Production cuts over to the new format. Since these two resource types are fairly stable, this can be done independently, but it must be done first.
2. (optional) The changes to the content validator to support the new formats is ready. NOTE: at this writing, on Translation Notes have a validation function; thus tc-create can move forward.

On a related note, since we do use CV on TN resources and they will also have a new TSV format, then that deployment will need to be coordinated.

**Comment by birchamp** [comment]
This is also waiting for the MD-TSV Converter to add TQ.

**Comment by birchamp** [comment]
@RobH123 TQ has been published in TSV now, correct?

**Comment by birchamp** [comment]
@RobH123 Can you publish TQ as TSV as long as the back converter is ready?

**Comment by RobH123** [comment]
Yip, TQ and OBS-TQ TSV7 are now published into CatalogNext.

**Comment by mandolyte** [comment]
@birchamp please close this issue. This is working in v1.2. Since the code switches based on filename patterns both TQ as markdown and TQ as TSV are supported at the same time.

**Comment by elsylambert** [comment]
![Screen Shot 2021-09-30 at 8.53.09 AM.png](https://images.zenhubusercontent.com/5d02efb2f425650e637abe10/1697bc92-e066-4fab-85d2-36156d193f00)
Not seeing the Question column in TSV in  1.3 branch. @mandolyte

**Comment by elsylambert** [comment]
Wrote up #1042 as a separate issue for column selection persistence.

**Comment by elsylambert** [comment]
@mandolyte From your notes/comments above "For TQ and SQ:

These TSV files have FIVE columns: Reference, ID, Tags, Question, Response.
"
I am seeing Two extra columns (making it 7 column TSV)- Quote and Occurrence  in TQs
![Screen Shot 2021-10-04 at 2.49.30 PM.png](https://images.zenhubusercontent.com/5d02efb2f425650e637abe10/ba61f301-6cd1-41c7-8ae9-286c3bd6c63e)
Is that correct?

**Comment by elsylambert** [comment]
@mandolyte Another problem I found is that markdown display ( upon Preview mode toggle ) is not working for the Question/Response columns.
"MD should be displayed the way TN displays but only display default column that are not empty."

**Comment by mandolyte** [comment]
@elsylambert  On the number of columns... that comment was made back in February and things have changed. If you check the actual repo [here](https://git.door43.org/unfoldingWord/en_tq/src/branch/master/tq_2JN.tsv)  you will see these have 7 columns. Not sure why, since the empty columns remain unused as far as I seen. But perhaps the content team has plans for them???

**Comment by elsylambert** [comment]
@mandolyte  Yes I saw the 7 column TQ in the repo and was confirming if that was right. Thanks for confirming.

**Comment by elsylambert** [comment]
translationCore Create - v1.4.0-rc.3 build 104-85b41fb.. Looks good. Tested with 7 column TSVs.
